sdk/search/Azure.Search.Documents/src/Batching/SearchIndexingBufferedSender{T}.cs

```diff
@@ -0,0 +1,606 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Azure.Search.Documents.Batching;
+using Azure.Search.Documents.Indexes;
+using Azure.Search.Documents.Indexes.Models;
+using Azure.Search.Documents.Models;
+
+namespace Azure.Search.Documents
+{
+    /// <summary>
+    /// Index search documents with intelligent batching, automatic flushing,
+    /// and retries for failed indexing actions.
+    /// </summary>
+    /// <typeparam name="T">
+    /// The .NET type that maps to the index schema.  Instances of this
+    /// type can be retrieved as documents from the index. You can use
+    /// <see cref="SearchDocument"/> for dynamic documents.
+    /// </typeparam>
+    public class SearchIndexingBufferedSender<T> : IDisposable, IAsyncDisposable
+    {
+        /// <summary>
+        /// Flag indicating whether the sender has been disposed.
+        /// </summary>
+        private volatile int _disposed = 0;
+
+        /// <summary>
+        /// The single publisher responsible for submitting requests.
+        /// </summary>
+        private SearchIndexingPublisher<T> _publisher;
+
+        /// <summary>
+        /// Gets the <see cref="SearchClient"/> used to send requests to the
+        /// service.
+        /// </summary>
+        internal virtual SearchClient SearchClient { get; }
+
+        /// <summary>
+        /// Gets the URI endpoint of the Search Service.  This is likely
+        /// to be similar to "https://{search_service}.search.windows.net".
+        /// </summary>
+        /// <remarks>
+        /// This is not the URI of the Search Index.  You could construct that
+        /// URI with "{Endpoint}/indexes/{IndexName}" if needed.
+        /// </remarks>
+        public virtual Uri Endpoint => SearchClient.Endpoint;
+
+        /// <summary>
+        /// Gets the name of the Search Service.
+        /// </summary>
+        public virtual string ServiceName => SearchClient.ServiceName;
+
+        /// <summary>
+        /// Gets the name of the Search Index.
+        /// </summary>
+        public virtual string IndexName => SearchClient.IndexName;
+
+        /// <summary>
+        /// Gets a function that can be used to access the index key value of a
+        /// document.
+        /// </summary>
+        internal virtual Func<T, string> KeyFieldAccessor { get; private set; }
+
+        /// <summary>
+        /// Ensure we only lookup the key field accessor once, but that
+        /// everyone waits until we have one.
+        /// </summary>
+        private InitializationBarrierSlim _keyFieldBarrier = new InitializationBarrierSlim();
```

Comment:
Is the requirement for everyone to wait required for the correctness of just an optimization?

Reply:
Because you can use `Lazy<Task>` if it's only for optimization.

Reply:
Required for correctness.  We can't process the results of a request without the KeyFieldAccessor.  We only want one thread to make the request if we have to though.

Reply:
Yes, but if two threads make it it's not the end of the world and won't impact the correctness, right? Doing one request is just an optimization.

Reply:
Some folks on our team would consider making an extra request a correctness bug, but I see what you mean.